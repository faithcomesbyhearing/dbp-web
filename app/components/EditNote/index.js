/**
 *
 * EditNote
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import PopupMessage from '../PopupMessage';
import SvgWrapper from '../SvgWrapper';
import NoteVerseText from '../NoteVerseText';

class EditNote extends React.PureComponent {
	constructor(props) {
		super(props);
		// TODO: Can get rid of the const statements below once the tags field is an object
		const hasTitle =
			this.props.note.get('tags') &&
			typeof this.props.note.get('tags').find === 'function' &&
			this.props.note.get('tags').find((t) => t.get('type') === 'title');
		const titleText =
			(hasTitle &&
				this.props.note
					.get('tags')
					.find((t) => t.get('type') === 'title')
					.get('value')) ||
			'';

		this.state = {
			textarea: this.props.note.get('notes') || '',
			savedNote: !!this.props.note.get('id'),
			selectedChapter: '',
			selectedBookName: '',
			selectedBookId: '',
			titleText,
			savingNote: false,
			readTheMessage: false,
			popupOpen: false,
			coords: { x: 0, y: 0 },
		};
	}

	componentDidUpdate() {
		if (this.state.savingNote && this.props.savedTheNote) {
			this.setState({ savingNote: false });
		}
	}

	getReference = (verseStart, verseEnd, chapter) =>
		`${this.props.activeBookName} ${chapter}:${
			verseStart === verseEnd || !verseEnd
				? verseStart
				: `${verseStart}-${verseEnd}`
		}`;

	getCurrentDate = () => {
		const date = new Date();
		const day = date.getDate();
		const year = date.getFullYear().toFixed().slice(2);
		const month = (date.getMonth() + 1).toFixed();

		return `${month.length === 1 ? `0${month}` : month}.${day}.${year}`;
	};

	get verseReference() {
		const { vernacularNamesObject, note } = this.props;
		if (
			note.get('tags') &&
			note.get('tags').find((tag) => tag.get('type') === 'reference')
		) {
			const ref = note
				.get('tags')
				.find((tag) => tag.get('type') === 'reference');
			return ref.get('value');
		}
		const book = note.get('book_id');
		const start = note.get('verse_start');
		const end = note.get('verse_end');
		const chapter = note.get('chapter');
		const verses = start === end || !end ? start : `${start}-${end}`;

		if (book && chapter && start) {
			return `${vernacularNamesObject[book]} ${chapter}:${verses}`;
		} else if (this.state.selectedBookName && this.state.selectedChapter) {
			return `${this.state.selectedBookName} ${this.state.selectedChapter}`;
		}
		return 'Please Add a Verse';
	}

	handleTextareaChange = (e) => {
		e.persist();
		if (
			!this.state.readTheMessage ||
			this.props.savedTheNote ||
			this.props.errorSavingNote
		) {
			this.props.readSavedMessage();
			this.props.clearNoteErrorMessage();
		}
		this.setState((cs) =>
			cs.readTheMessage
				? { textarea: e.target.value }
				: { textarea: e.target.value, readTheMessage: true },
		);
	};

	handleNoteTitleChange = (e) => {
		e.persist();
		if (
			!this.state.readTheMessage ||
			this.props.savedTheNote ||
			this.props.errorSavingNote
		) {
			this.props.readSavedMessage();
			this.props.clearNoteErrorMessage();
		}
		this.setState((cs) =>
			cs.readTheMessage
				? { titleText: e.target.value }
				: { titleText: e.target.value, readTheMessage: true },
		);
	};

	handleSave = (e) => {
		const { note, activeTextId, errorSavingNote } = this.props;
		const { titleText, textarea } = this.state;
		const chapter = note.get('chapter');
		const verseStart = note.get('verse_start');
		const verseEnd = note.get('verse_end');
		const bookId = note.get('book_id');
		const id = note.get('id');
		const hasTitle =
			this.props.note.get('tags') &&
			typeof this.props.note.get('tags').find === 'function' &&
			this.props.note.get('tags').find((t) => t.get('type') === 'title');
		const prevTitle =
			(hasTitle &&
				this.props.note
					.get('tags')
					.find((t) => t.get('type') === 'title')
					.get('value')) ||
			'';

		if (!textarea) {
			const rightEdge = window.innerWidth - 100;
			const bottomEdge = window.innerHeight - 115;
			const x = rightEdge < e.clientX ? rightEdge : e.clientX;
			const y = bottomEdge < e.clientY ? bottomEdge : e.clientY;

			this.setState({ popupOpen: true, coords: { x, y } });
			if (this.timer) {
				clearTimeout(this.timer);
			}
			this.timer = setTimeout(() => {
				// There may be an issue with doing this here because of the 2500 ms delay
				if (errorSavingNote) {
					this.props.clearNoteErrorMessage();
				}
				this.setState({ popupOpen: false });
			}, 2500);
			return;
		}

		if (note.get('notes') === textarea && prevTitle === titleText) {
			// If the original text equals the current text then I don't have to do anything
			// Because the user has not updated their note
			this.setState({ savingNote: false });
			return;
		} else if (this.state.savedNote) {
			this.props.updateNote({
				bible_id: activeTextId,
				notes: textarea,
				book_id: bookId,
				bookmark: 0,
				verse_start: verseStart,
				verse_end: verseEnd,
				chapter,
				tags: `title::: ${titleText},reference::: ${this.getReference(
					verseStart,
					verseEnd,
					chapter,
				)}`,
				id,
			});
		} else {
			this.props.addNote({
				bible_id: activeTextId,
				notes: textarea,
				book_id: bookId,
				bookmark: 0,
				verse_start: verseStart,
				verse_end: verseEnd,
				chapter,
				tags: `title::: ${titleText},reference::: ${this.getReference(
					verseStart,
					verseEnd,
					chapter,
				)}`,
			});
		}

		this.setState({ savingNote: true });
	};

	deleteNote = () => {
		if (this.props.note.get('id')) {
			this.props.deleteNote({ noteId: this.props.note.get('id') });
			this.props.setActiveChild('notes');
		}
	};

	render() {
		const {
			toggleVerseText,
			note,
			isVerseTextVisible,
			activeTextId,
			notePassage,
			savedTheNote,
			notesErrorMessage,
			errorSavingNote,
			loadingChapterForNote,
		} = this.props;
		const { savingNote, popupOpen, coords } = this.state;

		return (
			<section className="edit-notes">
				<div
					tabIndex={0}
					role={'button'}
					onClick={() => this.props.setActiveChild('notes')}
					className={'go-back'}
				>
					<SvgWrapper className={'icon'} svgid={'arrow_left'} />
					<span>View All Notes</span>
				</div>
				<div className="date-title">
					<input
						onChange={this.handleNoteTitleChange}
						placeholder="CLICK TO ADD TITLE"
						className="title"
						value={this.state.titleText}
					/>
					<span className="date">
						{note.get('date') || this.getCurrentDate()}
					</span>
				</div>
				<div className={`verse-dropdown${isVerseTextVisible ? ' open' : ''}`}>
					<SvgWrapper
						onClick={toggleVerseText}
						className={'icon'}
						svgid="arrow_right"
					/>
					<span className="text">{this.verseReference}</span>
					<span className="version-dropdown">
						{note.get('bible_id') || activeTextId}
					</span>
				</div>
				{isVerseTextVisible && (
					<NoteVerseText
						loading={loadingChapterForNote}
						notePassage={notePassage}
					/>
				)}
				<textarea
					onChange={this.handleTextareaChange}
					placeholder="CLICK TO ADD NOTE"
					value={this.state.textarea}
					className="note-text"
				/>
				<div className={'delete-note-section'}>
					<span
						className="delete-button"
						role="button"
						tabIndex={0}
						onClick={() => this.deleteNote()}
					>
						<SvgWrapper className={'icon'} svgid={'delete'} />
						Delete Note
					</span>
					{savedTheNote && !errorSavingNote ? (
						<span
							className="saved-note"
							role="button"
							tabIndex={0}
							onClick={this.handleSave}
						>
							Saved
						</span>
					) : (
						<span
							className={errorSavingNote ? 'save-button failed' : 'save-button'}
							role="button"
							tabIndex={0}
							onClick={this.handleSave}
						>
							{savingNote && !errorSavingNote
								? 'Updating...'
								: `${errorSavingNote ? 'Failed' : 'Save'}`}
						</span>
					)}
				</div>
				{popupOpen ? (
					<PopupMessage
						message={
							errorSavingNote && notesErrorMessage
								? notesErrorMessage
								: 'The notes field cannot be empty.'
						}
						x={coords.x}
						y={coords.y}
					/>
				) : null}
			</section>
		);
	}
}

EditNote.propTypes = {
	addNote: PropTypes.func,
	updateNote: PropTypes.func,
	deleteNote: PropTypes.func,
	setActiveChild: PropTypes.func,
	toggleVerseText: PropTypes.func,
	readSavedMessage: PropTypes.func,
	clearNoteErrorMessage: PropTypes.func,
	note: PropTypes.object,
	vernacularNamesObject: PropTypes.object,
	savedTheNote: PropTypes.bool,
	errorSavingNote: PropTypes.bool,
	isVerseTextVisible: PropTypes.bool,
	loadingChapterForNote: PropTypes.bool,
	notePassage: PropTypes.string,
	activeTextId: PropTypes.string,
	activeBookName: PropTypes.string,
	notesErrorMessage: PropTypes.string,
};

export default EditNote;
