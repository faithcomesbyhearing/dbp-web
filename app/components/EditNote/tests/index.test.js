import React from 'react';
import { render } from '@testing-library/react';
import EditNote from '..';

const props = {
	addNote: jest.fn(),
	updateNote: jest.fn(),
	deleteNote: jest.fn(),
	setActiveChild: jest.fn(),
	toggleVerseText: jest.fn(),
	readSavedMessage: jest.fn(),
	clearNoteErrorMessage: jest.fn(),
	note: structuredClone({
		notes: 'This is a test note!',
		bible_id: 'ENGESV',
		created_at: '2019-02-12 18:44:07',
		chapter: 3,
		book_id: 'ROM',
		verse_end: 23,
		updated_at: '2019-02-12 18:44:07',
		verse_start: 23,
		tags: [
			{
				id: 6,
				note_id: 403619,
				bookmark_id: null,
				highlight_id: null,
				type: 'title',
				value: 'My First Test Note',
			},
			{
				id: 7,
				note_id: 403619,
				bookmark_id: null,
				highlight_id: null,
				type: 'reference',
				value: 'Matthew 1:7',
			},
		],
		id: 403619,
	}),
	vernacularNamesObject: {},
	notePassage: 'For all have sinned and fallen short of the glory of God.',
	activeTextId: 'ENGESV',
	activeBookName: 'Romans',
	notesErrorMessage: '',
	isVerseTextVisible: true,
	savedTheNote: false,
	errorSavingNote: false,
};
const realDate = Date;
describe('EditNote component', () => {
	beforeAll(() => {
		global.Date = jest.fn(() => ({
			getDate: () => 20,
			getFullYear: () => 2019,
			getMonth: () => 1,
		}));
	});

	afterAll(() => {
		global.Date = realDate;
	});

	it('should match snapshot of edit note with default props', () => {
		const { asFragment } = render(<EditNote {...props} />);
		expect(asFragment()).toMatchSnapshot();
	});
	it('should match snapshot of edit note when verse text is not visible', () => {
		const { asFragment } = render(
			<EditNote {...props} isVerseTextVisible={false} />,
		);
		expect(asFragment()).toMatchSnapshot();
	});
	it('should match snapshot of edit note when it was saved successfully', () => {
		const { asFragment } = render(<EditNote {...props} savedTheNote />);
		expect(asFragment()).toMatchSnapshot();
	});
	it('should match snapshot of edit note when there was an error saving it', () => {
		const { asFragment } = render(<EditNote {...props} errorSavingNote />);
		expect(asFragment()).toMatchSnapshot();
	});
});
