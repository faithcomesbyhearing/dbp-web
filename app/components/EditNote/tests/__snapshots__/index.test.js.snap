// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditNote component should match snapshot of edit note when it was saved successfully 1`] = `
<DocumentFragment>
  <section
    class="edit-notes"
  >
    <div
      class="go-back"
      role="button"
      tabindex="0"
    >
      <svg
        class="icon"
      >
        <use
          xlink:href="/svglist.svg#arrow_left"
          xmlns:xlink="http://www.w3.org/1999/xlink"
        />
      </svg>
      <span>
        View All Notes
      </span>
    </div>
    <div
      class="date-title"
    >
      <input
        class="title"
        placeholder="CLICK TO ADD TITLE"
        value="My First Test Note"
      />
      <span
        class="date"
      >
        02.20.19
      </span>
    </div>
    <div
      class="verse-dropdown open"
    >
      <svg
        class="icon"
      >
        <use
          xlink:href="/svglist.svg#arrow_right"
          xmlns:xlink="http://www.w3.org/1999/xlink"
        />
      </svg>
      <span
        class="text"
      >
        Matthew 1:7
      </span>
      <span
        class="version-dropdown"
      >
        ENGESV
      </span>
    </div>
    <div
      class="verse-text"
    >
      " For all have sinned and fallen short of the glory of God. "
    </div>
    <textarea
      class="note-text"
      placeholder="CLICK TO ADD NOTE"
    >
      This is a test note!
    </textarea>
    <div
      class="delete-note-section"
    >
      <span
        class="delete-button"
        role="button"
        tabindex="0"
      >
        <svg
          class="icon"
        >
          <use
            xlink:href="/svglist.svg#delete"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          />
        </svg>
        Delete Note
      </span>
      <span
        class="saved-note"
        role="button"
        tabindex="0"
      >
        Saved
      </span>
    </div>
  </section>
</DocumentFragment>
`;

exports[`EditNote component should match snapshot of edit note when there was an error saving it 1`] = `
<DocumentFragment>
  <section
    class="edit-notes"
  >
    <div
      class="go-back"
      role="button"
      tabindex="0"
    >
      <svg
        class="icon"
      >
        <use
          xlink:href="/svglist.svg#arrow_left"
          xmlns:xlink="http://www.w3.org/1999/xlink"
        />
      </svg>
      <span>
        View All Notes
      </span>
    </div>
    <div
      class="date-title"
    >
      <input
        class="title"
        placeholder="CLICK TO ADD TITLE"
        value="My First Test Note"
      />
      <span
        class="date"
      >
        02.20.19
      </span>
    </div>
    <div
      class="verse-dropdown open"
    >
      <svg
        class="icon"
      >
        <use
          xlink:href="/svglist.svg#arrow_right"
          xmlns:xlink="http://www.w3.org/1999/xlink"
        />
      </svg>
      <span
        class="text"
      >
        Matthew 1:7
      </span>
      <span
        class="version-dropdown"
      >
        ENGESV
      </span>
    </div>
    <div
      class="verse-text"
    >
      " For all have sinned and fallen short of the glory of God. "
    </div>
    <textarea
      class="note-text"
      placeholder="CLICK TO ADD NOTE"
    >
      This is a test note!
    </textarea>
    <div
      class="delete-note-section"
    >
      <span
        class="delete-button"
        role="button"
        tabindex="0"
      >
        <svg
          class="icon"
        >
          <use
            xlink:href="/svglist.svg#delete"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          />
        </svg>
        Delete Note
      </span>
      <span
        class="save-button failed"
        role="button"
        tabindex="0"
      >
        Failed
      </span>
    </div>
  </section>
</DocumentFragment>
`;

exports[`EditNote component should match snapshot of edit note when verse text is not visible 1`] = `
<DocumentFragment>
  <section
    class="edit-notes"
  >
    <div
      class="go-back"
      role="button"
      tabindex="0"
    >
      <svg
        class="icon"
      >
        <use
          xlink:href="/svglist.svg#arrow_left"
          xmlns:xlink="http://www.w3.org/1999/xlink"
        />
      </svg>
      <span>
        View All Notes
      </span>
    </div>
    <div
      class="date-title"
    >
      <input
        class="title"
        placeholder="CLICK TO ADD TITLE"
        value="My First Test Note"
      />
      <span
        class="date"
      >
        02.20.19
      </span>
    </div>
    <div
      class="verse-dropdown"
    >
      <svg
        class="icon"
      >
        <use
          xlink:href="/svglist.svg#arrow_right"
          xmlns:xlink="http://www.w3.org/1999/xlink"
        />
      </svg>
      <span
        class="text"
      >
        Matthew 1:7
      </span>
      <span
        class="version-dropdown"
      >
        ENGESV
      </span>
    </div>
    <textarea
      class="note-text"
      placeholder="CLICK TO ADD NOTE"
    >
      This is a test note!
    </textarea>
    <div
      class="delete-note-section"
    >
      <span
        class="delete-button"
        role="button"
        tabindex="0"
      >
        <svg
          class="icon"
        >
          <use
            xlink:href="/svglist.svg#delete"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          />
        </svg>
        Delete Note
      </span>
      <span
        class="save-button"
        role="button"
        tabindex="0"
      >
        Save
      </span>
    </div>
  </section>
</DocumentFragment>
`;

exports[`EditNote component should match snapshot of edit note with default props 1`] = `
<DocumentFragment>
  <section
    class="edit-notes"
  >
    <div
      class="go-back"
      role="button"
      tabindex="0"
    >
      <svg
        class="icon"
      >
        <use
          xlink:href="/svglist.svg#arrow_left"
          xmlns:xlink="http://www.w3.org/1999/xlink"
        />
      </svg>
      <span>
        View All Notes
      </span>
    </div>
    <div
      class="date-title"
    >
      <input
        class="title"
        placeholder="CLICK TO ADD TITLE"
        value="My First Test Note"
      />
      <span
        class="date"
      >
        02.20.19
      </span>
    </div>
    <div
      class="verse-dropdown open"
    >
      <svg
        class="icon"
      >
        <use
          xlink:href="/svglist.svg#arrow_right"
          xmlns:xlink="http://www.w3.org/1999/xlink"
        />
      </svg>
      <span
        class="text"
      >
        Matthew 1:7
      </span>
      <span
        class="version-dropdown"
      >
        ENGESV
      </span>
    </div>
    <div
      class="verse-text"
    >
      " For all have sinned and fallen short of the glory of God. "
    </div>
    <textarea
      class="note-text"
      placeholder="CLICK TO ADD NOTE"
    >
      This is a test note!
    </textarea>
    <div
      class="delete-note-section"
    >
      <span
        class="delete-button"
        role="button"
        tabindex="0"
      >
        <svg
          class="icon"
        >
          <use
            xlink:href="/svglist.svg#delete"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          />
        </svg>
        Delete Note
      </span>
      <span
        class="save-button"
        role="button"
        tabindex="0"
      >
        Save
      </span>
    </div>
  </section>
</DocumentFragment>
`;
